apps/ensrainbow/src/utils/env-utils.ts

```diff
@@ -0,0 +1,298 @@
+      );
+    }
+
+    // if a validator is provided, run it
+    if (validator) {
+      const validationResult = validator(rawValue);
+      if (validationResult !== true) {
+        const errorMessage =
+          typeof validationResult === "string"
+            ? validationResult
+            : `Invalid value '${rawValue}' for environment variable '${envVarName}'.`;
+        throw new Error(errorMessage);
+      }
+    }
+
+    return rawValue;
+  } catch (error) {
+    const errorMessage = formatEnvVarError(envVarName, error);
+    throw new Error(errorMessage);
+  }
+};
+
+/**
+ * Parses a non-negative integer from an environment variable
+ *
+ * @param envVarName The name of the environment variable
+ * @param defaultValue Optional default value if the environment variable is not set
+ * @returns The parsed non-negative integer
+ * @throws Error if the value is not a valid non-negative integer
+ */
+export const getEnvNonNegativeInteger = (envVarName: string, defaultValue?: number): number => {
+  try {
+    const rawValue = process.env[envVarName];
+
+    // if no value is provided, return the default or throw an error
+    if (rawValue === undefined) {
+      if (defaultValue !== undefined) {
+        return defaultValue;
+      }
+      throw new Error(
+        `Environment variable '${envVarName}' is not set and no default value was provided.`,
+      );
+    }
+
+    try {
+      return parseNonNegativeInteger(rawValue);
+    } catch (error: unknown) {
+      if (error instanceof Error) {
+        throw new Error(`Invalid value for environment variable '${envVarName}': ${error.message}`);
+      }
+      throw error;
+    }
+  } catch (error) {
+    const errorMessage = formatEnvVarError(envVarName, error);
+    throw new Error(errorMessage);
+  }
+};
+
+/**
+ * Parses a file path from an environment variable and optionally validates that the file exists and is readable
+ *
+ * @param envVarName The name of the environment variable
+ * @param defaultValue Optional default value if the environment variable is not set
+ * @param should_exist Whether to check if the file exists (default: true)
+ * @param should_be_readable Whether to check if the file is readable (default: true)
+ * @returns The parsed file path
+ * @throws Error if the file does not exist (when should_exist is true) or is not readable (when should_be_readable is true)
+ */
+export const getEnvFilePath = (
+  envVarName: string,
+  defaultValue?: string,
+  should_exist: boolean = true,
+  should_be_readable: boolean = true,
+): string => {
+  try {
+    const filePath = getEnvString(envVarName, defaultValue);
+
+    try {
+      return parseFilePath(filePath, should_exist, should_be_readable);
+    } catch (error) {
+      throw new Error(`${error instanceof Error ? error.message : String(error)}`);
+    }
+  } catch (error) {
+    const errorMessage = formatEnvVarError(envVarName, error);
+    throw new Error(errorMessage);
+  }
```

Comment:
`formatEnvVarError` handles error type conversion to string, so we could flatten the try/catch structure:
```suggestion
  try {
    const filePath = getEnvString(envVarName, defaultValue);

    return parseFilePath(filePath, should_exist, should_be_readable);
  } catch (error) {
    const errorMessage = formatEnvVarError(envVarName, error);
    throw new Error(errorMessage);
  }
```